### PR TITLE
doOnEach - reduce allocations, report to RxJavaHooks

### PR DESCRIPTION
**Pull request by davidmoten**
as discussed in #4260 
- call `RxJavaHooks.onError` for secondary errors
- reduced allocations (`ActionSubscriber` -> `ActionObserver`, `Operator` -> `OnSubscribe`)
- added 5 unit tests

### CODE REVIEW

File: src/main/java/rx/Observable.java [hunk 1]
```diff
@@ -5509,7 +5509,7 @@ public Completable toCompletable() {
         Action1<Throwable> onError = Actions.empty();
         Observer<T> observer = new ActionSubscriber<T>(onNext, onError, onCompleted);
```
Comment (akarnokd) [inline src/main/java/rx/Observable.java]: ActionObserver?

File: src/test/java/rx/internal/operators/OnSubscribeReduceTest.java [hunk 2]
```diff
@@ -273,5 +273,9 @@ public Integer call(Integer a, Integer b) {
         ts.assertError(NoSuchElementException.class);
     }
 
-
+    @Test
+    public void test() {
+
```
Comment (akarnokd) [inline src/test/java/rx/internal/operators/OnSubscribeReduceTest.java]: Empty test?

File: src/main/java/rx/Observable.java (hunk 1 above)
Comment (davidmoten) [inline src/main/java/rx/Observable.java]: ta

**Comment by davidmoten** [comment]
fixed ActionSubscriber usage and removed empty test, ta

**Comment by akarnokd** [comment]
:+1:

**Pull request merged**